### ISSUE TITLE
TimerNode should be able to show units

### ISSUE THREAD

**Issue by samreid** [phetsims/scenery-phet#372]
In Wave Interference, we would like to show time in femtoseconds for the light scene.  However, the timer node looks like this and has no place for units.  

![image](https://user-images.githubusercontent.com/679486/39199261-3fc2b336-47a6-11e8-8042-80b914f74393.png)

This will require design work and implementation work.  @arouinfar can you please comment how you would like this to appear with units?

**Comment by arouinfar** [comment]
@samreid we'll probably want to do something like this:

![image](https://user-images.githubusercontent.com/8419308/39442721-72f03986-4c6f-11e8-9c60-5586bbcecba9.png)

I expanded the width of the body by 15% to accommodate the extra characters needed for the units. I left the buttons in place because I think it looks strange to keep them aligned with the edges of the readout.
![image](https://user-images.githubusercontent.com/8419308/39442697-63694f5c-4c6f-11e8-9563-75111c613c48.png)

What do you think @samreid? Since this would be a change to a common component, it would be good for @ariel-phet to also review.

**Comment by ariel-phet** [comment]
@arouinfar - I agree with keeping the button positions

@samreid I think you would want two options 
1. Where no units are displayed keeping the smaller form factor. I think in general it is fine not to display units when the timer is in seconds
2. Where units are displayed and using the larger timer body.

**Comment by samreid** [comment]
One question about how to deal with units for different scenes.  The strings "ms" and "fs" have different widths--should the timer node resize when we change scenes?  Or should we try to choose the max width across strings so that it doesn't change size when changing scenes?

**Comment by samreid** [comment]
Some questions I've asked on the developer channel:

- [ ] I’m adding features to SCENERY_PHET/TimerNode.  Are we going all in on the “stop propagating subtype options to the supertype”?
- [ ] I’m struggling with making TimerNode have mutable units-- it is based on RectangularButtonView which has a fixed minWidth (cannot be changed after construction).  Should I update that too?  Or should I just create different TimerNodes with different fixed units?  I thought we decided we preferred mutable components, but it is going to be a ton of work to get there.

**Comment by samreid** [comment]
We should sort out the design questions in https://github.com/phetsims/scenery-phet/issues/372#issuecomment-385567732 and https://github.com/phetsims/scenery-phet/issues/372#issuecomment-385716388 then ask the development team how to proceed on https://github.com/phetsims/scenery-phet/issues/372#issuecomment-385732672.  I'll stash my work somehow and wait for discussion.

**Comment by samreid** [comment]
Here's a poor man's branch of TimerNode.js, collapsed so it doesn't clog up this issue:

<details><summary>Poor Man's Branch of TimerNode.js</summary>
<p>

```js
// Copyright 2014-2018, University of Colorado Boulder

/**
 * Timer Node.
 *
 * @author Jonathan Olson <jonathan.olson@colorado.edu>
 * @author Anton Ulyanov (Mlearner)
 */
define( function( require ) {
  'use strict';

  // modules
  var BooleanRectangularToggleButton = require( 'SUN/buttons/BooleanRectangularToggleButton' );
  var Bounds2 = require( 'DOT/Bounds2' );
  var inherit = require( 'PHET_CORE/inherit' );
  var Node = require( 'SCENERY/nodes/Node' );
  var Path = require( 'SCENERY/nodes/Path' );
  var PhetFont = require( 'SCENERY_PHET/PhetFont' );
  var Rectangle = require( 'SCENERY/nodes/Rectangle' );
  var RectangularPushButton = require( 'SUN/buttons/RectangularPushButton' );
  var sceneryPhet = require( 'SCENERY_PHET/sceneryPhet' );
  var ShadedRectangle = require( 'SCENERY_PHET/ShadedRectangle' );
  var Shape = require( 'KITE/Shape' );
  var Tandem = require( 'TANDEM/Tandem' );
  var Text = require( 'SCENERY/nodes/Text' );
  var Util = require( 'DOT/Util' );
  var UTurnArrowShape = require( 'SCENERY_PHET/UTurnArrowShape' );

  /**
   *
   * @param {Property.<number>} secondsProperty
   * @param {Property.<boolean>} runningProperty
   * @param {Object} [options]
   * @constructor
   */
  function TimerNode( secondsProperty, runningProperty, options ) {
    Tandem.indicateUninstrumentedCode();
    options = _.extend( {
      iconColor: '#333',
      buttonBaseColor: '#DFE0E1',
      touchAreaDilation: 10,
      units: 'fs',
      cursor: 'pointer'
    }, options );

    Node.call( this );

    /*---------------------------------------------------------------------------*
     * Readout text
     *----------------------------------------------------------------------------*/
    var largeNumberText = new PhetFont( 20 );
    var bigReadoutText = new Text( timeToBigString( 0 ), {
      font: largeNumberText
    } );
    var smallFont = new PhetFont( 15 );
    var smallReadoutText = new Text( timeToSmallString( 0 ), {
      font: smallFont,
      left: bigReadoutText.right
    } );
    this.unitsText = new Text( options.units || '', {
      font: smallFont,
      left: smallReadoutText.right + 4
    } );
    // aligns the baselines of the big and small text
    smallReadoutText.bottom = smallReadoutText.bounds.maxY - bigReadoutText.bounds.minY;
    bigReadoutText.top = 0;
    this.unitsText.bottom = smallReadoutText.bottom;
    var readoutText = new Node( {
      children: [
        bigReadoutText,
        smallReadoutText,
        this.unitsText
      ],
      pickable: false
    } );
    readoutText.centerX = 0;

    /*---------------------------------------------------------------------------*
     * Readout background
     *----------------------------------------------------------------------------*/
    var textBackground = Rectangle.roundedBounds( readoutText.bounds.dilatedXY( 5, 2 ), 5, 5, {
      fill: '#fff',
      stroke: 'rgba(0,0,0,0.5)',
      pickable: false
    } );

    var paddingBetweenItems = 6;
    var minimumButtonWidth = ( textBackground.width - paddingBetweenItems ) / 2 - 1; // -1 due to the stroke making it look mis-aligned

    /*---------------------------------------------------------------------------*
     * Buttons
     *----------------------------------------------------------------------------*/
    var resetAllShape = new UTurnArrowShape( 10 );
    var playPauseHeight = resetAllShape.bounds.height;
    var playPauseWidth = playPauseHeight;
    var halfPlayStroke = 0.05 * playPauseWidth;
    var playOffset = 0.15 * playPauseWidth;
    var playShape = new Shape().moveTo( playPauseWidth - halfPlayStroke * 0.5 - playOffset, 0 )
      .lineTo( halfPlayStroke * 1.5 + playOffset, playPauseHeight / 2 - halfPlayStroke - playOffset )
      .lineTo( halfPlayStroke * 1.5 + playOffset, -playPauseHeight / 2 + halfPlayStroke + playOffset )
      .close().getOffsetShape( -playOffset );
    // a stop symbol (square)
    var pauseShape = Shape.bounds( new Bounds2( 0, -playPauseHeight / 2, playPauseWidth, playPauseHeight / 2 ).eroded( playPauseWidth * 0.1 ) );

    var resetButton = new RectangularPushButton( {
      listener: function resetTimer() {
        runningProperty.set( false );
        secondsProperty.set( 0 );
      },
      content: new Path( resetAllShape, {
        fill: options.iconColor
      } ),
      baseColor: options.buttonBaseColor,
      minWidth: minimumButtonWidth
    } );

    var playPauseButton = new BooleanRectangularToggleButton(
      new Path( pauseShape, { fill: options.iconColor } ),
      new Path( playShape, {
        stroke: options.iconColor,
        fill: '#eef',
        lineWidth: halfPlayStroke * 2
      } ), runningProperty, {
        baseColor: options.buttonBaseColor,
        minWidth: minimumButtonWidth
      } );

    /*---------------------------------------------------------------------------*
     * Layout
     *----------------------------------------------------------------------------*/
    var container = new Node();
    container.addChild( resetButton );
    container.addChild( playPauseButton );
    container.addChild( textBackground );
    container.addChild( readoutText );

    resetButton.right = -paddingBetweenItems / 2;
    playPauseButton.left = paddingBetweenItems / 2;
    resetButton.top = textBackground.bottom + paddingBetweenItems;
    playPauseButton.top = textBackground.bottom + paddingBetweenItems;

    var panelPad = 8;
    container.left = panelPad;
    container.top = panelPad;

    /*---------------------------------------------------------------------------*
     * Panel background
     *----------------------------------------------------------------------------*/
    var roundedRectangle = new ShadedRectangle( container.bounds.dilated( panelPad ) );
    roundedRectangle.touchArea = roundedRectangle.localBounds.dilated( options.touchAreaDilation );
    this.addChild( roundedRectangle );
    this.addChild( container );

    /*---------------------------------------------------------------------------*
     * Control logic
     *----------------------------------------------------------------------------*/
    var updateTime = function updateTime( value ) {
      bigReadoutText.text = timeToBigString( value );
      smallReadoutText.text = timeToSmallString( value );
      resetButton.enabled = value > 0;
    };
    secondsProperty.link( updateTime );

    /*---------------------------------------------------------------------------*
     * Target for drag listeners
     *----------------------------------------------------------------------------*/
    this.dragTarget = roundedRectangle;

    // @private
    this.disposeTimerNode = function() {
      secondsProperty.unlink( updateTime );
      resetButton.dispose();
      playPauseButton.dispose();
    };

    this.mutate( options );

    // @private
    this.updateTextBackgroundShape = function() {
      textBackground.setRectBounds( readoutText.bounds.dilatedXY( 5, 2 ) );
    };
  }

  sceneryPhet.register( 'TimerNode', TimerNode );

  // the full-sized minutes and seconds string
  function timeToBigString( timeInSeconds ) {
    // Round to the nearest centisecond (compatible with timeToSmallString).
    // see https://github.com/phetsims/masses-and-springs/issues/156
    timeInSeconds = Util.roundSymmetric( timeInSeconds * 100 ) / 100;

    var minutes = Math.floor( timeInSeconds / 60 ) % 60;
    var seconds = Math.floor( timeInSeconds ) % 60;

    if ( seconds < 10 ) {
      seconds = '0' + seconds;
    }
    if ( minutes < 10 ) {
      minutes = '0' + minutes;
    }
    return minutes + ':' + seconds;
  }

  // the smaller hundredths-of-a-second string
  function timeToSmallString( timeInSeconds ) {
    // Round to the nearest centisecond (compatible with timeToSmallString).
    // see https://github.com/phetsims/masses-and-springs/issues/156
    timeInSeconds = Util.roundSymmetric( timeInSeconds * 100 ) / 100;

    // Rounding after mod, in case there is floating-point error
    var centiseconds = Util.roundSymmetric( timeInSeconds % 1 * 100 );
    if ( centiseconds < 10 ) {
      centiseconds = '0' + centiseconds;
    }
    return '.' + centiseconds;
  }

  return inherit( Node, TimerNode, {
    // @public - Provide dispose() on the prototype for ease of subclassing.
    dispose: function() {
      this.disposeTimerNode();
    },

    /**
     * Set the units for the readout.
     * @public
     * @param {string|null} units
     */
    setUnits: function( units ) {
      this.units = units;
      this.unitsText.text = units || '';
      this.updateTextBackgroundShape();
    }
  } );
} );
```

</p>
</details>

**Comment by arouinfar** [comment]
>  I'd expect to see the edges of the buttons to line up with the edges of the readout on the left and right side whether or not units are shown. This can be accomplished without a big open spot (trapped space) in the middle by making the buttons a little wider for the case with units.

Great suggestion @samreid! Here's what it looks like when we make the buttons a bit wider:
![image](https://user-images.githubusercontent.com/8419308/39488409-6ab2a0f8-4d3f-11e8-9674-0c9bb13a24cb.png)

> I agree generally we can suppress the "s" readout when it is supposed to match real-time seconds. However in sims where there are multiple time scales, we may wish to show the "s" for clarity. 

Agreed. Sims with multiple time scales should always show a unit, even when "s". 

>The strings "ms" and "fs" have different widths--should the timer node resize when we change scenes? Or should we try to choose the max width across strings so that it doesn't change size when changing scenes?

Great question. Let's pick a max width that accommodates the longest string, so TimerNode doesn't resize between scenes.

**Comment by ariel-phet** [comment]
Agree with @arouinfar comments above.

**Comment by samreid** [comment]
Also, one question for the reviewer: should we generalize it to be able to use arbitrary Nodes for units, or is it simpler/better to require that the units are strings?

**Comment by ariel-phet** [comment]
@jonathanolson said he is available for review

**Comment by jonathanolson** [comment]
> I’m adding features to SCENERY_PHET/TimerNode. Are we going all in on the “stop propagating subtype options to the supertype”?

It's probably good to do that.

> I’m struggling with making TimerNode have mutable units-- it is based on RectangularButtonView which has a fixed minWidth (cannot be changed after construction). Should I update that too? Or should I just create different TimerNodes with different fixed units? I thought we decided we preferred mutable components, but it is going to be a ton of work to get there.

MutableOptionsNode sounds helpful here (for now). I agree mutable components are the long-term best solution.

> Also, one question for the reviewer/designers: should we generalize it to be able to use arbitrary Nodes for units, or is it simpler/better to require that the units are strings?

RichText may be sufficient (think m^2), but we may need arbitrary nodes for some things.

Re the code:

How is layout supposed to work? Right now if you have one "longer" unit and one "shorter" one (horizontally), the unit is left-aligned always (resulting in a lot of padding on the right for the shorter one). e.g.:

<img width="195" alt="screen shot 2018-05-02 at 9 43 19 pm" src="https://user-images.githubusercontent.com/480878/39559233-e47184be-4e51-11e8-8a65-bb0a5769c8de.png">

Can we just adaptively update all of the layout based on the current units? (e.g. just pass in a unitsProperty). That would presumably simplify things, no?

It also felt a bit un-natural to default to the first units (instead of null) if unitsChoices is provided (with no initial units). Also iterating through the units Texts checking their .text would be simplified if we just use unitsProperty.

Is there a specific reason the mutate is done before the setUnits call? Usually we'd fully construct the object before the mutate.

Thoughts?

**Comment by samreid** [comment]
At yesterday's design meeting, we decided the text should be centered in the panel, but the panel shouldn't resize when different units are specified.  We decided to use text, noting that this will predominantly be used for time measures and hence strings like `m^2` will not be used, though we pointed out one day we may be tempted to use something like `10^-15 sec`, but we are hoping not.

That being said, I like the idea of using a unitsProperty, or perhaps unitsNodeProperty.

**Comment by samreid** [comment]
> I’m adding features to SCENERY_PHET/TimerNode. Are we going all in on the “stop propagating subtype options to the supertype”?

The PhET Software Design Patterns document now says:

> We do not filter child options out before passing them up to the parent. With this in mind please be mindful of the option naming to make sure that you don't cause collisions. See https://github.com/phetsims/tasks/issues/934.

After seeing how complicated it was to separate supertype options/subtype options and omit them, I'm going to move back to the "pass everything" pattern.

**Comment by samreid** [comment]
I considered going with a Property for the units node, but realized we would need to know the possible values to allocate the correct amount of space for it.  I didn't want to expose and rely on Property.validValues, so I decided to investigate using ToggleNode for that part.  That leads to a clean implementation, but still has the layout problem identified in https://github.com/phetsims/scenery-phet/issues/372#issuecomment-386185029 . I'll commit what I have so far, but will need to continue on this.

**Comment by samreid** [comment]
The preceding commit has better support for showing units, but revealed we should factor out a type for the readouts (specifically one base type and two subtypes), so that there is less scattered logic in TimerNode itself.  The TimerNode seems to be behaving OK now in Wave Interference and Wave on a String, but I've marked this as blocks-publication until the work is completed and reviewed.

**Comment by samreid** [comment]
I factored out TimerReadoutNode and it is simple enough that I don't think we need subclasses or inversion of control.  These changes appear to be OK in Wave Interference, Wave on a String and the Scenery-PhET test harness.  These changes should be reviewed before we publish any simulations that use it.  @jonathanolson you are most familiar with TimerNode, can you please review it and TimerReadoutNode?

**Comment by zepumph** [comment]
Discussed in dev meeting today. We think that @Denz1994 would be a good candidate for a review of this. Assigning to him.

**Comment by pixelzoom** [comment]
I don't see the need for TimerReadoutNode.

TimerNode should have these options

- `valueFormat: '{{value}}'`.  If you want ms units, then pass in `valueFormat: '{{value}} ms'`.
- `valueAlign: 'right'`. How the value is aligned in the display field, 'left'|'right'|'center'.  
- `buttonXSpacing: 5`.  Change this if you want more/less space between the buttons. 

Use RichText for the value display.

**Comment by samreid** [comment]
@pixelzoom said:

> If this can display femtoseconds or milliseconds, the property shouldn't be called secondsProperty

**Comment by ariel-phet** [comment]
@samreid can you make a separate issue for instrumenting for iO with you and @Denz1994 as assignee

**Comment by samreid** [comment]
> @samreid can you make a separate issue for instrumenting for iO with you and @Denz1994 as assignee

Will do.  I thought @jonathanolson referred to something, perhaps a CT breakage from this change, I'll take a look and create issue as appropriate.

**Comment by samreid** [comment]
@pixelzoom suggested the units may need to precede the value depending on i18

**Comment by samreid** [comment]
Capacitor Lab: Basics is awaiting redeploy from master, and could use this change.

**Comment by Denz1994** [comment]
Once Masses and Springs is published we will revisit the state of PhET-IO with @ariel-phet during a design meeting.

**Comment by samreid** [comment]
Remaining work is tracked in issues mentioned above.  Since TimerNode can now show units, I'll close this issue and we can follow up in the remaining open issues.
